Quantum-Control-Applications/Superconducting/Two-Flux-Tunable-Coupled-Transmons/Use Case 4 - Single- and Two-Qubit State and Process Tomography/two-qubit-state-tomography.py

```diff
@@ -0,0 +1,349 @@
+# "q<qubit>_z", and readout resonator elements "rr<qubit>"
+
+qubit1 = 1
+qubit2 = 2
+
+if qubit1 == 1:
+    threshold1 = ge_threshold_q1
+elif qubit1 == 2:
+    threshold1 = ge_threshold_q2
+else:
+    raise ValueError(f"Incorrect qubit1 number chosen")
+
+if qubit2 == 1:
+    threshold2 = ge_threshold_q1
+elif qubit2 == 2:
+    threshold2 = ge_threshold_q2
+else:
+    raise ValueError(f"Incorrect qubit2 number chosen")
+
+if qubit1 == qubit2:
+    raise ValueError(f"The value of qubit1 cannot equal that of qubit2")
+
+
+n_avg = 10_000
+
+
+# subroutine to prepare desired qubit state
+def prepare_state(qubit1, qubit2):
+    # write whatever QUA code you need in order to create the
+    # state to perform tomography on, from an initial
+    # ground state. For example, to create the |1> 1/sqrt(2)(|0>+i|1>)
+    # state
+    play("y180", f"q{qubit1}_xy")
+    play("-x90", f"q{qubit2}_xy")
+
+
+# matrix representation of the ideal composite qubit state from above
+# (|1> 1/sqrt(2)(|0>+i|1>)) (<1| 1/sqrt(2)(<0|-i<1|))
+ideal_state = np.array([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 1 / 2, 1j / 2], [0, 0, -1j / 2, 1 / 2]])
+
+
+with program() as two_qubit_state_tomography:
+
+    n = declare(int)
+    n_st = declare_stream()
+    I = [declare(fixed) for _ in range(2)]
+    Q = [declare(fixed) for _ in range(2)]
+    # I_st = [declare_stream() for _ in range(2)]
+    # Q_st = [declare_stream() for _ in range(2)]
+
+    states = [
+        declare(bool),
+        declare(bool),
+    ]  # QUA variable for the measured qubit states
+    # states_st = [declare_stream(), declare_stream()]  # Stream for the qubits states
+
+    p00 = declare(int)  # variable to track number of |0>|0> measurements
+    p01 = declare(int)  # variable to track number of |0>|1> measurements
+    p10 = declare(int)  # variable to track number of |1>|0> measurements
+    p11 = declare(int)  # variable to track number of |1>|1> measurements
+
+    prob_vec_results_st = declare_stream()  # Stream for the probability vector - average of states vector
+
+    c = declare(int)  # QUA variable for switching between projections
+
+    with for_(n, 0, n < n_avg, n + 1):  # QUA for_ loop for averaging
+        with for_(c, 1, c <= 15, c + 1):  # QUA for_ loop for switching between projections
+
+            prepare_state(qubit1, qubit2)
+            align()
+
+            assign(p00, 0)
+            assign(p01, 0)
+            assign(p10, 0)
+            assign(p11, 0)
+
+            with switch_(c):
+                with case_(1):  # projection along Z1, X2
+
+                    play("-y90", f"q{qubit2}_xy")
+                    # Stokes parameter for this case: P00 + P10 - P01 - P11
+
+                with case_(2):  # projection along Z1, Y2
+
+                    play("x90", f"q{qubit2}_xy")
+                    # Stokes parameter for this case: P00 + P10 - P01 - P11
+
+                with case_(3):  # projection along Z1, Z2
+
+                    pass
```

Comment:
Hi @bornman-nick this is very interesting, I would have thought the program would not compiled because a `pass` in python is not a `pass` in QUA, so with case_(3) its an empty line of QUA code. I know you are trying to do a ZZ, so to keep the same length of the switch cases how about to use identity gate as in `wait(pi_len)`

Reply:
Note this would fail (or more correctly, probably produce an unexpected output) if the flag `unsafe` is set to `True`.